File: ios/GradeThread/Grading/ConsumerGradeFlow.swift
import Foundation
import Observation

/// US-2016 — the consumer grade journey as a state machine.
///
/// submit -> pay -> poll -> result, with two states that must say the money
/// part FIRST. The walk-around screen already learned that lesson and its
/// comment says why: "we couldn't grade it" reads like a wasted purchase until
/// you know it was not one. An abstain and a credits prompt are both
/// no-charge states, and both are commonly mistaken for failures.
///
/// A STATE MACHINE RATHER THAN A VIEW WITH FLAGS, because the flow has a real
/// order and two of its steps can bounce back: paying can send you to buy
/// credits and return, and an abstain sends you back to the camera. Booleans
/// for "isSubmitting / isPaying / isPolling" make those legal transitions look
/// like edge cases.
@MainActor
@Observable
final class ConsumerGradeFlow {
    enum Step: Equatable {
        case ready
        /// Bytes moving. `fraction` is the upload only.
        case uploading(fraction: Double)
        /// Uploaded; charging against included grades or credits.
        case paying(submissionId: String)
        /// Neither covered it. NOT an error - an offer, and the pack is the one
        /// the route named.
        case needsCredits(submissionId: String, offer: PhotoGradePayment.PackOffer?)
        /// The purchase went through and the server has not credited the account
        /// YET.
        ///
        /// ⚠ THIS STATE IS THE WHOLE POINT OF THE TOP-UP FLOW. An Apple purchase
        /// completing on the device does not mean the balance has moved: the
        /// grant arrives through the server, and there is a gap. Paying again
        /// during that gap returns "out of credits" a SECOND time - the same
        /// wall the customer just paid to clear.
        case awaitingCredits(submissionId: String)
        /// The grant did not appear inside the poll window. NOT a failure and
        /// NOT a refusal: it may still land, so the user gets "check again"
        /// rather than an error.
        case creditsDelayed(submissionId: String)
        /// Paid, grading. The server sends nothing until it is done, so the UI
        /// must render this indeterminately rather than invent a number.
        case grading(submissionId: String, statusText: String)
        /// The gate abstained. Nothing was charged.
        case needsPhotos(submissionId: String, messages: [String], slots: [String])
        case graded(submissionId: String)
        case failed(String)
    }

    private(set) var step: Step = .ready

    // `@escaping` on the progress callback, because that is what the thing it is
    // handed to requires: `PhotoGradeUploadService.submit` takes
    // `@MainActor @escaping (Double) -> Void` and holds it across the upload.
    // A closure parameter of a closure TYPE is non-escaping by default, so
    // without this the default `submit` below cannot forward its own argument.
    private let submit: ([PhotoGradeImage], PhotoGradeRequest, @MainActor @escaping (Double) -> Void) async throws -> PhotoGradeOutcome
    private let pay: (String) async throws -> PhotoGradePayment.Outcome
    private let status: (String) async throws -> PhotoGradeStatus
    private let sleep: (Double) async -> Void

    /// THE SAME top-up flow ``GradeRequestStore`` uses, not a second one.
    ///
    /// It polls the billing snapshot until the balance rises above the
    /// pre-purchase baseline, tolerating a failed read, and it re-checks once
    /// more on timeout because the grant may land between the last poll and
    /// giving up. Two answers to "have the credits arrived yet" would drift, and
    /// the one that drifts is the one that tells a paying customer no.
    let creditTopUp: CreditTopUpFlow

    /// Seams injected so the whole journey is testable without a network. The
    /// defaults are the real calls.
    init(
        submit: (([PhotoGradeImage], PhotoGradeRequest, @MainActor @escaping (Double) -> Void) async throws -> PhotoGradeOutcome)? = nil,
        pay: ((String) async throws -> PhotoGradePayment.Outcome)? = nil,
        status: ((String) async throws -> PhotoGradeStatus)? = nil,
        sleep: ((Double) async -> Void)? = nil,
        creditTopUp: CreditTopUpFlow? = nil
    ) {
        // Built in the init body, not as a default argument: a default is
        // evaluated in a nonisolated context and PaywallStore is @MainActor -
        // the same reason GradeRequestStore constructs its own here.
        self.creditTopUp = creditTopUp ?? CreditTopUpFlow(
            fetchBalance: { await PaywallStore.liveBillingFetcher()?.credits })
        self.submit = submit ?? { images, request, onProgress in
            try await PhotoGradeUploadService.submit(
                images: images, request: request, onProgress: onProgress)
        }
        self.pay = pay ?? { submissionId in
            let response: PhotoGradePayResponse = try await EdgeAPI.shared.postJSON(
                "/api/grade/pay/\(submissionId)",
                body: PhotoGradePayRequest(tier: "standard"))
            return response.outcome()
        }
        self.status = status ?? { submissionId in
            try await EdgeAPI.shared.getJSON("/api/grade/status/\(submissionId)")
        }
        self.sleep = sleep ?? { seconds in
            try? await Task.sleep(for: .seconds(seconds))
        }
    }

    func start(images: [PhotoGradeImage], request: PhotoGradeRequest) async {
        step = .uploading(fraction: 0)
        let submissionId: String
        do {
            let outcome = try await submit(images, request) { [weak self] fraction in
                guard let self else { return }
                // Never backwards, never past 1. A retried body segment can
                // report a lower cumulative count, and a bar that jumps back
                // reads as a failure.
                if case let .uploading(current) = self.step {
                    self.step = .uploading(fraction: max(current, min(1, fraction)))
                }
            }
            guard case let .submitted(id, alreadyPaid) = outcome else { return }
            submissionId = id
            if alreadyPaid {
                await poll(submissionId: submissionId)
                return
            }
        } catch {
            step = .failed(message(for: error))
            return
        }
        await charge(submissionId: submissionId)
    }

    /// Called when the credit purchase completes.
    ///
    /// ⚠ WAITS FOR THE GRANT BEFORE RETRYING, which my first version did not.
    /// The purchase succeeding on the device is not the balance moving; the
    /// grant comes through the server and there is a gap. Charging inside that
    /// gap answers "out of credits" again and shows the customer the same wall
    /// they just paid to clear.
    ///
    /// `baseline` is the balance BEFORE the purchase. The grant is detected as a
    /// strict increase over it, which tolerates a failed read and does not
    /// assume the pack size arrived exactly.
    ///
    /// Retrying the charge itself is safe: the route enforces one debit per
    /// submission in the database (US-2298, after batch grading charged up to
    /// five times for one garment).
    func creditsPurchased(submissionId: String, baseline: Int) async {
        step = .awaitingCredits(submissionId: submissionId)
        await creditTopUp.awaitGrant(baseline: baseline) { [weak self] in
            guard let self else { return }
            if case .granted = self.creditTopUp.state {
                await self.charge(submissionId: submissionId)
            }
        }
        // timedOut is not a refusal - the grant may still land, so offer
        // "check again" rather than an error. awaitGrant already re-checks once
        // more before reporting it.
        if case .timedOut = creditTopUp.state, case .awaitingCredits = step {
            step = .creditsDelayed(submissionId: submissionId)
        }
    }

    /// The "check again" affordance on ``Step/creditsDelayed``.
    func recheckCredits(submissionId: String, baseline: Int) async {
        await creditsPurchased(submissionId: submissionId, baseline: baseline)
    }

    private func charge(submissionId: String) async {
        step = .paying(submissionId: submissionId)
        do {
            switch try await pay(submissionId) {
            case .paidFromIncluded, .paidFromCredits:
                await poll(submissionId: submissionId)
            case let .needsCredits(offer):
                step = .needsCredits(submissionId: submissionId, offer: offer)
            }
        } catch {
            step = .failed(message(for: error))
        }
    }

    private func poll(submissionId: String) async {
        var attempt = 0
        step = .grading(submissionId: submissionId, statusText: "Grading your photos…")
        while !Task.isCancelled {
            do {
                let current = try await status(submissionId)
                if current.isTerminal {
                    step = terminalStep(for: current, submissionId: submissionId)
                    return
                }
                step = .grading(
                    submissionId: submissionId,
                    statusText: PhotoGradeStatusRules.waitingText(for: current.status))
            } catch {
                // A transient read failure is not a failed grade. The grade is
                // running server-side either way, so keep polling rather than
                // telling the user their submission died.
                step = .grading(
                    submissionId: submissionId,
                    statusText: "Still working…")
            }
            await sleep(PhotoGradeStatusRules.nextDelaySeconds(attempt: attempt))
            attempt += 1
        }
    }

    private func terminalStep(for status: PhotoGradeStatus, submissionId: String) -> Step {
        if status.isGraded { return .graded(submissionId: submissionId) }
        if status.needsPhotos {
            return .needsPhotos(
                submissionId: submissionId,
                messages: PhotoGradeStatusRules.retakeMessages(status.quality_feedback),
                slots: PhotoGradeStatusRules.blockingSlots(status.quality_feedback))
        }
        // expired / failed / disputed. Named rather than swallowed into
        // "something went wrong", because each has a different next move.
        return .failed(ConsumerGradeCopy.terminalMessage(for: status.status))
    }

    private func message(for error: Error) -> String {
        (error as? LocalizedError)?.errorDescription
            ?? "We couldn't send that for grading."
    }
}

struct PhotoGradePayRequest: Encodable {
    let tier: String
}

enum ConsumerGradeCopy {
    /// A terminal status that is not a grade and not an abstain.
    static func terminalMessage(for status: String) -> String {
        switch status {
        case "expired":
            return "This grade expired before it was paid for. Start it again when you're ready."
        case "failed":
            return "Grading didn't finish. You weren't charged - try again, and tell support if it repeats."
        case "disputed":
            return "This grade is under dispute. We'll update you when it's reviewed."
        default:
            return "That grade didn't finish."
        }
    }

    /// The headline for a no-charge state. Said BEFORE the reason, because
    /// "we couldn't grade it" reads like a wasted purchase until you know it
    /// was not one - the walk-around screen's own lesson.
    static let noChargeHeadline = "You weren't charged for this."
}

File: android/app/src/main/java/com/gradethread/app/grading/ConsumerGradeFlow.kt
package com.gradethread.app.grading

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * US-2815: the consumer grade journey as a state machine.
 *
 * submit -> pay -> poll -> result, with two states that must say the MONEY part
 * first. The walk-around screen learned that lesson first and its comment says
 * why: "we couldn't grade it" reads like a wasted purchase until you know it
 * was not one. An abstain and a credits prompt are both no-charge states, and
 * both are commonly mistaken for failures.
 *
 * A STATE MACHINE RATHER THAN A SCREEN WITH FLAGS, because the flow has a real
 * order and two of its steps bounce back: paying can send you to buy credits and
 * return, and an abstain sends you back to the camera. Booleans for
 * isSubmitting / isPaying / isPolling make those legal transitions look like
 * edge cases.
 *
 * Every collaborator is a function parameter with a real default. That is what
 * lets the whole journey — including the credits detour and the poll timeout —
 * be tested without a network, a clock, or a store.
 */
class ConsumerGradeFlow(
    private val submit: suspend (List<PhotoGradeImage>, PhotoGradeRequest, (Double) -> Unit) -> PhotoSubmitResponse,
    private val pay: suspend (String) -> PhotoGradePayment.Outcome,
    private val status: suspend (String) -> PhotoGradeStatus,
    private val sleep: suspend (Long) -> Unit,
) {

    sealed class Step {
        object Ready : Step()

        /** Bytes moving. [fraction] is the upload only. */
        data class Uploading(val fraction: Double) : Step()

        /** Uploaded; charging against included grades or credits. */
        data class Paying(val submissionId: String) : Step()

        /**
         * Neither covered it. NOT an error — an offer, and the pack is the one
         * the route named.
         */
        data class NeedsCredits(
            val submissionId: String,
            val offer: PhotoGradePayment.PackOffer?,
        ) : Step()

        /**
         * The purchase went through and the server has not credited the account
         * YET.
         *
         * ⚠ THIS STATE IS THE WHOLE POINT OF THE TOP-UP DETOUR. A store purchase
         * completing on the device does not mean the balance has moved: the
         * grant arrives through the server, and there is a gap. Paying again
         * during that gap returns "out of credits" a SECOND time — the same wall
         * the customer just paid to clear.
         */
        data class AwaitingCredits(val submissionId: String) : Step()

        /**
         * The grant did not appear inside the poll window. NOT a failure and NOT
         * a refusal: it may still land, so the user gets "check again" rather
         * than an error.
         */
        data class CreditsDelayed(val submissionId: String) : Step()

        /**
         * Paid, grading. The server sends nothing until it is done, so the UI
         * must render this indeterminately rather than invent a number.
         */
        data class Grading(val submissionId: String, val statusText: String) : Step()

        /** The gate abstained. Nothing was charged. */
        data class NeedsPhotos(
            val submissionId: String,
            val messages: List<String>,
            val slots: List<String>,
        ) : Step()

        data class Graded(val submissionId: String) : Step()

        data class Failed(val message: String) : Step()
    }

    private val stepFlow = MutableStateFlow<Step>(Step.Ready)
    val step: StateFlow<Step> = stepFlow.asStateFlow()

    suspend fun start(images: List<PhotoGradeImage>, request: PhotoGradeRequest) {
        // Refuse here rather than after the upload: the route's abstain costs a
        // charge and a vision call per image before the refund.
        PhotoGradeUploader.validate(images)?.let {
            stepFlow.value = Step.Failed(it.message)
            return
        }

        stepFlow.value = Step.Uploading(0.0)
        val submissionId = try {
            val response = submit(images, request) { fraction ->
                stepFlow.value = Step.Uploading(fraction)
            }
            response.submissionId
        } catch (error: Throwable) {
            stepFlow.value = Step.Failed(message(error))
            return
        }

        if (submissionId.isNullOrEmpty()) {
            stepFlow.value = Step.Failed("That upload did not complete. Try again.")
            return
        }
        charge(submissionId)
    }

    /** After the top-up sheet reports a grant. */
    suspend fun creditsPurchased(submissionId: String) {
        stepFlow.value = Step.AwaitingCredits(submissionId)
        charge(submissionId)
    }

    /** The "check again" affordance behind [Step.CreditsDelayed]. */
    suspend fun recheckCredits(submissionId: String) = charge(submissionId)

    private suspend fun charge(submissionId: String) {
        stepFlow.value = Step.Paying(submissionId)
        val outcome = try {
            pay(submissionId)
        } catch (error: Throwable) {
            stepFlow.value = Step.Failed(message(error))
            return
        }
        when (outcome) {
            is PhotoGradePayment.Outcome.NeedsCredits ->
                stepFlow.value = Step.NeedsCredits(submissionId, outcome.offer)
            else -> poll(submissionId)
        }
    }

    private suspend fun poll(submissionId: String) {
        stepFlow.value = Step.Grading(submissionId, "")
        for (attempt in 0 until MAX_POLLS) {
            val current = try {
                status(submissionId)
            } catch (error: Throwable) {
                stepFlow.value = Step.Failed(message(error))
                return
            }
            if (current.isTerminal) {
                stepFlow.value = terminalStep(current, submissionId)
                return
            }
            stepFlow.value = Step.Grading(
                submissionId,
                PhotoGradeStatus.terminalMessage(current.status),
            )
            if (attempt < MAX_POLLS - 1) sleep(POLL_DELAY_MS)
        }
        // Ran out of polls. NOT a failure: the grade is still running and the
        // certificate will exist. Saying "failed" here would be a lie that
        // invites a second charge.
        stepFlow.value = Step.Grading(submissionId, "Still grading. This one is taking a while.")
    }

    private fun terminalStep(status: PhotoGradeStatus, submissionId: String): Step =
        when (status.status) {
            "completed" -> Step.Graded(submissionId)
            "needs_photos" -> Step.NeedsPhotos(
                submissionId,
                // The gate's own wording. Re-deriving copy from `issues` would
                // be building worse sentences out of better data.
                status.qualityFeedback?.photoRequests
                    ?: listOfNotNull(status.qualityFeedback?.summary),
                status.qualityFeedback?.photoSlots ?: emptyList(),
            )
            else -> Step.Failed(PhotoGradeStatus.terminalMessage(status.status))
        }

    private fun message(error: Throwable): String =
        error.message?.takeIf { it.isNotBlank() } ?: "Something went wrong. Try again."

    companion object {
        /** ~2 minutes at 4s, which covers a normal grade with room to spare. */
        const val MAX_POLLS = 30
        const val POLL_DELAY_MS = 4_000L
    }
}
